### PR TITLE
Register tools as a reloadSources service

### PR DESCRIPTION
**Pull request by B3rn475**
In https://github.com/dart-lang/sdk/commit/df8bf384eb815cf38450cb50a0f4b62230fba217 a new functionality of the Dart VM Service Protocol has been introduced.

Clients connected to the Service Protocol are now able to expose services that other clients (e.g. Observatory) can invoke through the Service Protocol itself.

With these changes Flutter Tools register them self as a `reloadSources` (a.k.a. HotReload) capable client.
Observatory is already listening for the clients which expose this functionality and uses by default the service based version of  `reloadSources` when available, so requesting a HotReload from Observatory will trigger the full Flutter HotReload.

Related https://github.com/dart-lang/sdk/issues/30023

### CODE REVIEW

**Comment by B3rn475** [comment]
@jason-simmons @tvolkert

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -44,12 +44,15 @@ class FlutterDevice {
     _viewsCache = null;
   }
 
-  void connect() {
+  /// If the [reloadSources] parameter is not null the 'reloadSources' service
+  /// will be registered
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/resident_runner.dart]: nit: end comments with a period

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 2]
```diff
@@ -84,13 +84,20 @@ class HotRunner extends ResidentRunner {
     return true;
   }
 
+  Future _reloadSourcesService(String isolateId,
+      {bool force: false, bool pause: false}) async {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/run_hot.dart]: nit: spaces around named arguments

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 3]
```diff
@@ -24,6 +24,10 @@ typedef StreamChannel<String> _OpenChannel(Uri uri);
 
 _OpenChannel _openChannel = _defaultOpenChannel;
 
+/// A function that reacts to the invocation of the 'reloadSources' service
+typedef Future ReloadSources(String isolateId,
+    {bool force, bool pause});
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/vmservice.dart]: ditto

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 4]
```diff
@@ -40,13 +44,50 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
     _vm = new VM._empty(this);
     _peer.listen().catchError(_connectionError.completeError);
 
     _peer.registerMethod('streamNotify', (rpc.Parameters event) {
       _handleStreamNotify(event.asMap);
     });
+
+    if (reloadSources != null) {
+      _peer.registerMethod('reloadSources', (rpc.Parameters params) async {
+        const kInvalidParams = -32602;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/vmservice.dart]: missing types

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 5]
```diff
@@ -40,13 +44,50 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
+
+        if (isolateId is! String || isolateId == '')
+          throw new rpc.RpcException.invalidParams('Invalid \'isolateId\'');
+        if (force is! bool)
+          throw new rpc.RpcException.invalidParams('Invalid \'force\'');
+        if (force is! bool)
+          throw new rpc.RpcException.invalidParams('Invalid \'pause\'');
+
+        try {
+          await reloadSources(isolateId, force: force ?? false,
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/vmservice.dart]: maybe the arguments should

Comment (Hixie) [inline packages/flutter_tools/lib/src/vmservice.dart]: aren't force and pause guaranteed to be non-null here?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 6]
```diff
@@ -78,14 +119,17 @@ class VMService {
   ///
   /// Requests made via the returns [VMService] time out after [requestTimeout]
   /// amount of time, which is [kDefaultRequestTimeout] by default.
+  /// If the [reloadSources] parameter is not null the 'reloadSources' service
+  /// will be registered
   static VMService connect(
     Uri httpUri, {
     Duration requestTimeout: kDefaultRequestTimeout,
+    ReloadSources reloadSources
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/vmservice.dart]: nit: trailing comma

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Could you elaborate on this documentation? Suppose someone doesn't know what reloadSources are. The current documentation tells them very little.

Comment (B3rn475) [inline packages/flutter_tools/lib/src/resident_runner.dart]: done

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 2 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/run_hot.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 3 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 4 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 5 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: You are right, i forgot to remove them during a refactoring.

done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 6 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/resident_runner.dart]: I've added a more detailed description of the process and of the reasons around it.

done

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 7]
```diff
@@ -24,6 +24,10 @@ typedef StreamChannel<String> _OpenChannel(Uri uri);
 
 _OpenChannel _openChannel = _defaultOpenChannel;
 
+/// A function that reacts to the invocation of the 'reloadSources' service.
+typedef Future ReloadSources(String isolateId,
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Nit: format

```
typedef Future ReloadSources(
  String isolateId, {
  bool force,
  bool pause.
});
```

Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Seems like this is `Future<Null>` (not `Future<dynamic>`).  Either way, please update it to be explicit.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 8]
```diff
@@ -84,13 +84,20 @@ class HotRunner extends ResidentRunner {
     return true;
   }
 
+  Future _reloadSourcesService(String isolateId,
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/run_hot.dart]: Per comment below, I think you want this to be `Future<Null>`

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 9]
```diff
@@ -78,14 +118,17 @@ class VMService {
   ///
   /// Requests made via the returns [VMService] time out after [requestTimeout]
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: s/returns/returned/

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 10]
```diff
@@ -78,14 +118,17 @@ class VMService {
...
   /// amount of time, which is [kDefaultRequestTimeout] by default.
+  /// If the [reloadSources] parameter is not null the 'reloadSources' service
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Nit: add blank comment line before this.

Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: It's not clear from this comment what it means for the 'reloadSources' service to be registered.  Can you elaborate?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 11]
```diff
@@ -24,6 +24,10 @@ typedef StreamChannel<String> _OpenChannel(Uri uri);
 
 _OpenChannel _openChannel = _defaultOpenChannel;
 
+/// A function that reacts to the invocation of the 'reloadSources' service.
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Can you explain what the 'reloadSources' service is here?

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 10 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Nit: comma after "is not null"

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 12]
```diff
@@ -40,13 +44,49 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
 
 /// A connection to the Dart VM Service.
 class VMService {
-  VMService._(this._peer, this.httpAddress, this.wsAddress, this._requestTimeout) {
+  VMService._(this._peer, this.httpAddress, this.wsAddress, this._requestTimeout,
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Can you put each parameter on its own line now?

```
VMService._(
  this._peer,
  this.httpAddress,
  this.wsAddress,
  this._requestTimeout,
  ReloadSources reloadSources,
) {
```

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 13]
```diff
@@ -40,13 +44,49 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
+    if (reloadSources != null) {
+      _peer.registerMethod('reloadSources', (rpc.Parameters params) async {
+        const int kInvalidParams = -32602;
+        const int kServerError = -32000;
+
+        final String isolateId = params['isolateId'].value;
+        final bool force = params.asMap['force'] ?? false;
+        final bool pause = params.asMap['pause'] ?? false;
+
+        if (isolateId is! String || isolateId == '')
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: s/ == ''/.isEmpty/

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 14]
```diff
@@ -40,13 +44,49 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
+          throw new rpc.RpcException.invalidParams('Invalid \'isolateId\'');
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Add the value of the param that was received to the exception message for these checks.

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 15]
```diff
@@ -40,13 +44,49 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
     _vm = new VM._empty(this);
     _peer.listen().catchError(_connectionError.completeError);
 
     _peer.registerMethod('streamNotify', (rpc.Parameters event) {
       _handleStreamNotify(event.asMap);
     });
+
+    if (reloadSources != null) {
+      _peer.registerMethod('reloadSources', (rpc.Parameters params) async {
+        const int kInvalidParams = -32602;
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Remove this (it's unused)

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 16]
```diff
@@ -40,13 +44,49 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
+        const int kServerError = -32000;
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: Just use `rpc_error_code.SERVER_ERROR` -- we're already importing `'package:json_rpc_2/error_code.dart' as rpc_error_code`

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 8 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/run_hot.dart]: `Future<Null>` is probably better, I'll also add an exception to allow the service to return a failure message

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 10 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 9 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 14 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 13 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 16 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 7 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 12 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 15 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 11 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 10 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 17]
```diff
@@ -40,13 +57,51 @@ const Duration kShortRequestTimeout = const Duration(seconds: 5);
...
+          throw new rpc.RpcException.invalidParams('Invalid \'pause\': $pause');
+
+        try {
+          await reloadSources(isolateId, force: force, pause: pause);
+          return {'type': 'Success'};
+        } on rpc.RpcException {
+          rethrow;
+        } catch (e, st) {
+          throw new rpc.RpcException(rpc_error_code.SERVER_ERROR,
+              'Error during Sources Reload: $e$st');
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/vmservice.dart]: `\n` between `$e` and `$st`

**Review by tvolkert — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 17 above)
Comment (B3rn475) [inline packages/flutter_tools/lib/src/vmservice.dart]: done

**Comment by zanderso** [comment]
It looks like this change introduced these analysis errors

```
Analyzing /usr/local/google/home/zra/fuchsia/fuchsia/lib/flutter/packages/flutter_tools/lib...
  error • Place 'package:' imports before relative imports at package:flutter_tools/src/run_hot.dart:19:1 • directives_ordering
  error • Place 'package:' imports before relative imports at package:flutter_tools/src/run_hot.dart:20:1 • directives_ordering
  error • Avoid empty statements at package:flutter_tools/src/run_hot.dart:97:6 • empty_statements
  error • Specify type annotations at package:flutter_tools/src/vmservice.dart:89:18 • always_specify_types
  error • Specify type annotations at package:flutter_tools/src/vmservice.dart:100:50 • always_specify_types
```

**Pull request merged**